---
title: "Transcriptome Analysis"
output: html_document
---

```{r setup, include=FALSE}
library(DESeq2)
library(ggplot2)
library(tidyverse)
library(pheatmap)

setwd("C:/Users/jnguy/Desktop/Burkett/DEG/")
knitr::opts_chunk$set(echo = TRUE)
```

#Differential Expression of Genes, requires DESeq2 and ggplot2
#https://r.omics.wiki/statistics/deseq2/example# 
#https://www.youtube.com/watch?v=wPzeea1Do18#
```{r}
design <- read_csv("C:/Users/Jen/Desktop/Burkett_Lab/R/Transcriptome/design.csv") 

cts <- read_csv("C:/Users/Jen/Desktop/Burkett_Lab/R/Transcriptome/counts.csv") %>% filter(Symbol != '') %>% distinct(Symbol, .keep_all = TRUE)

dds <- DESeqDataSetFromMatrix(countData = cts %>% column_to_rownames(var = "Symbol"), 
                              colData = design %>% column_to_rownames(var = "Sample"),
                              design = ~ Group)

keep <- rowSums(fpm(dds) >= 5) >= 1 * ncol(dds) 
dds <- dds[keep,]

dds <- estimateSizeFactors(dds)

ddsNormalized <- counts(dds, normalized=TRUE)

pheatmap(ddsNormalized,
         scale="row",
         cluster_cols=T,
         show_rownames = F,
         annotation_col = design %>% column_to_rownames(var = "Sample"),
         clustering_method = "ward.D2",
         main = "ETY311",
         color=colorRampPalette(c("blue","white","red"))(256))

dds <- DESeq(dds)

res <- results(dds, 
        contrast=c("Group", "Group_1", "Group_2"),
        independentFiltering = TRUE,
        alpha = 0.05) %>% as_tibble(rownames = "Gene.name")

hist(res$pvalue)
table(res$pvalue <= .05)

write_csv(res, "Group_2.Group_1.csv")
```

```{r}
head(res, 20)
```




#GSEA
#https://www.youtube.com/watch?v=Bzu4_yDcBLY#
```{r}
library(tidyverse)
library(msigdbr)
library(clusterProfiler)
library(fgsea)
model.results<-read_csv("C:/Users/jnguy/Desktop/Burkett/GSEA/model.results.csv")
```
##Enrichment
#Get gene set database
#https://igordot.github.io/msigdbr/articles/msigdbr-intro.html# 

```{r}
kegg<-msigdbr(species = "mouse", category="C2", subcategory = "CP:KEGG")
class(kegg)

```

##define sig genes
```{r}
ggplot(model.results, aes(x=padj)) + geom_histogram(bins=100) + theme_classic()
```
##set cutoff(padj same as fdr)
```{r}
signif<-model.results%>%
  filter(padj<=0.7)
```
##selects only column needed
```{r}
signif.gene<-unique(signif$gene_symbol)
keggdata<-select(kegg, gs_name, gene_symbol)
```

##run enrichment
```{r}
enrich.kegg<-enricher(gene = signif.gene, TERM2GENE = keggdata)
```

##extract results
```{r}
enrichresult<-enrich.kegg@result

```
#enrichR with wikipathways
```{r}
model.results2<-read_csv("C:/Users/jnguy/Desktop/Burkett/GSEA/model.results.csv")
wiki<-msigdbr(species = "mouse", category="C2", subcategory = "CP:WIKIPATHWAYS")
class(wiki)
signif.gene<-unique(signif$gene_symbol)
wikidata<-select(wiki, gs_name, gene_symbol)
enrich.wiki<-enricher(gene = signif.gene, TERM2GENE = wikidata)
wikiresults<-enrich.wiki@result
```
#X2K Transcription to Kinases

#UKA

#KRSA kinome
#devtools::install_github("CogDisResLab/KRSA")#
#https://kalganem.shinyapps.io/KRSA/#
```{r}
library(KRSA)
library(knitr)
library(tidyverse)
library(gt) 
library(furrr)
```
# Define chip type
```{r}
chipType <- "STK"
```
# Read crosstab view bionavigator files and tidy data
```{r}
setwd("C:\\Users\\jnguy\\Desktop\\Burkett\\KRSA")

data <- krsa_read("_ExportstosDBstsCopy_Median_SigmBg_201007153530.txt", "_ExportstosDBstsCopy_Signal_Saturation_201007153531.txt")
```


```{r}
krsa_qc_steps(data) -> data
data %>% mutate(Group = SampleName, SampleName = paste0(SampleName,"_" ,Barcode)) -> data

# extract end level signal values @ all max exposure time (200ms)
krsa_extractEndPointMaxExp(data, chipType) -> data_pw_200

# extract end level signal values end level @ all exposure times
krsa_extractEndPoint(data, chipType) -> data_pw 
```

QC steps and model fitting

```{r}
# Filter out peptides with low signals
krsa_filter_lowPeps(data_pw_200, 5) -> ppPassAll
#> Filtered out 79 Peptides

# Fit the linear model
# This will produce a list of data frames:
# scaled: the Slope_Transformed values (see above for more info on Slope_Transformed)
# normalized: the Slope_Transformed values but normalized by Chip/Barcode
# Grouped: The mean of Slope_Transformed values for all samples within a group
krsa_scaleModel(data_pw, ppPassAll) -> data_modeled
  
# Filter out peptides weak linear fit
krsa_filter_nonLinear(data_modeled$scaled, 0.8) -> ppPassR2
#> Filtered out 0 Peptides

# Filter out reference peptides
krsa_filter_ref_pep(ppPassR2) -> new_pep
#> Filtered out 10 Peptides
```
Plot global intensity - does not work
```{r}
krsa_cv_plot(data_modeled$scaled, new_pep)
```
Plot cv - does not work
```{r}
krsa_cv_plot(data_modeled$normalized, new_pep)
```
Global Volin Plots - sort of works
```{r}
krsa_violin_plot(data_modeled$scaled, new_pep, "Group")
```

Violin Plot
```{r}
krsa_violin_plot(data_modeled$scaled, new_pep, "Barcode")
```
Global Heatmap 

```{r}
krsa_heatmap(data_modeled$scaled, new_pep, scale = "row")
```
Global Heatmap with normalized Data

```{r}
krsa_heatmap(data_modeled$normalized, new_pep, scale = "row")
```
Generates model using the modeled group data

```{r}
krsa_heatmap_grouped(data_modeled$grouped, new_pep, scale = "row")
```
Making Comparsions 

```{r}
# Define Groups to be compared
comparisons <- list(Comp1 = c("A1", "A2", "A3", "A4", "B1" ,"B2", "B3", "B4", "C1", "C2", "C3", "C4", "D1", "D2", "D3", "D4"))

# This function will run both QC steps (krsa_filter_lowPeps, krsa_filter_nonLinear) and krsa_filter_ref_pep
krsa_quick_filter(data = data_pw_200, data2 = data_modeled$scaled,
    signal_threshold = 5, r2_threshold = 0.9,
    groups = comparisons$Comp1) -> pep_passed_qc
#> Filtered out 65 Peptides
#> Filtered out 1 Peptides
#> Filtered out 10 Peptides

# This function calculates 2 fold change values between the defined groups
# The byChip argument lets you calculates the log2 fold change the results within each chip
krsa_group_diff(data_modeled$scaled, comparisons$Comp1, pep_passed_qc, byChip = T) -> diff_df

# save LFC table
#write_delim(diff_df, "comp1_LFC_df.txt", delim = "\t")

# Extract top peptides based on the LFC cutoff using average of LFCs across chips
krsa_get_diff(diff_df,totalMeanLFC ,c(0.2,0.3)) %>% list("meanLFC" = .) -> sigPeps

# Extract top peptides based on the LFC cutoff using average of LFCs across chips with a standard deviation cutoff (sd_thr). in this example the sd_thr is set as 0.5 but that could be adjusted by user (uncomment next line to account for LFC sd)
# krsa_get_diff(diff_df,totalMeanLFC ,c(0.2,0.3), sd_thr = 0.5) %>% list("meanLFC" = .) -> sigPeps

# Extract top peptides based on the LFC cutoff using per chip LFCs
krsa_get_diff_byChip(diff_df, LFC, c(0.2,0.3)) -> sigPepsPerChip

# Combine the peptides hits in one list
sigPeps_total <- list(sigPeps, sigPepsPerChip) %>% unlist(recursive = F) %>%  unlist(recursive = F)
```
Heatmap

```{r}
krsa_heatmap(data_modeled$normalized, sigPeps$meanLFC$`0.2`, groups = comparisons$Comp1, scale = "row")
```

Comaprsion Violin plot

```{r}
krsa_violin_plot(data_modeled$scaled, sigPeps$meanLFC$`0.2`, "Barcode", groups = comparisons$Comp1)
```
KRSA violin plot grouped

```{r}
krsa_violin_plot_grouped(data_modeled$scaled, sigPeps$meanLFC$`0.2`, comparisons, 
                         dots = F,
                         groups = comparisons$Comp1, avg_line = T)
```
KRSA violin plot grouped

```{r}
krsa_violin_plot_grouped(data_modeled$grouped, sigPeps$meanLFC$`0.2`, comparisons, 
                         test = F, 
                         groups = comparisons$Comp1, avg_line = T)
```
Waterfall plot

```{r}
krsa_waterfall(diff_df, lfc_thr = 0.2, byChip = T)
```
```{r}
krsa_waterfall(diff_df, lfc_thr = 0.2, byChip = T, sd_thr = 0.5)
```
Curve plot 

```{r}
krsa_curve_plot(data_pw, sigPeps$meanLFC$`0.2`[1:10], groups = comparisons$Comp1)
```
```{r}
chipCov <- KRSA_coverage_STK_PamChip_87102_v1
KRSA_file <- KRSA_Mapping_STK_PamChip_87102_v1



# run the KRSA function to do the random sampling analysis, set seed that can be used later to reproduce results, and choose number of iterations
krsa(sigPeps$meanLFC$`0.2`, return_count = T, seed = 123, itr = 2000,
     map_file = KRSA_file, cov_file = chipCov) -> fin
```

# View the Z score table
```{r}
kable(head(fin$KRSA_Table,100), digits = 3)
```


Upstream analysis

```{r}
# to save file
fin$KRSA_Table %>% write_delim("acrossChip_KRSA_FullTable_comp1.txt", delim = "\t")

# find top and bottom kinases
bothways <- c(pull(head(fin$KRSA_Table, 10), Kinase), pull(tail(fin$KRSA_Table, 10), Kinase))

# Use these kinase to generate histogram plots for each selected kinase
krsa_histogram_plot(fin$KRSA_Table, fin$count_mtx, bothways)
```

```{r}
# For parallel computing, load the furrr package:
# opens multiple R sessions to run faster
plan(multisession)

# Run the KRSA function across the different sets of peptides using the furrr package for parallel computing
future_map(sigPeps_total, krsa) -> mutiple_krsa_outputs


# For none parallel computing:
# Run KRSA function across the different sets of peptides
#map(sigPeps_total, krsa) -> mutiple_krsa_outputs

# Tidy output
df <- data.frame(matrix(unlist(mutiple_krsa_outputs), ncol = max(lengths(mutiple_krsa_outputs)), byrow = TRUE))
df <- setNames(do.call(rbind.data.frame, mutiple_krsa_outputs), names(mutiple_krsa_outputs$meanLFC.0.2))

df %>% rownames_to_column("method") %>% select(Kinase, Z, method) %>% 
  mutate(method = str_extract(method, "\\w+\\.\\w+\\.\\w+")) %>% 
  mutate(method = gsub("(^\\w+)[\\.]", "\\1>", method)) %>% 
  mutate_if(is.numeric, round, 2) -> df

df %>% 
  pivot_wider(names_from = method, values_from = Z) -> df2
  
# Generates a table of the Z scores across the different sets of peptides
# df2 %>% 
#   gt() %>% tab_spanner_delim(delim = ">")

# Creates an average Z score table using the across chip analysis
df %>% 
  filter(grepl("mean", method)) %>% 
  select(Kinase, Z, method) %>% group_by(Kinase) %>% mutate(AvgZ = mean(Z)) -> AvgZTable

# save file
#AvgZTable %>% write_delim("acrossChip_KRSA_Table_comp1.txt", delim = "\t")

# Creates an average Z score table using the within chip analysis
df %>% 
  filter(!grepl("mean", method)) %>% 
  select(Kinase, Z, method) %>% group_by(Kinase) %>% mutate(AvgZ = mean(Z)) -> AvgZTable2

# save file
#AvgZTable %>% write_delim("withinChip_KRSA_Table_comp1.txt", delim = "\t")

# Extract top kinases based on abs(Z) score
krsa_top_hits(AvgZTable2, 2) -> kinases_hits

#krsa_top_hits(AvgZTable2, 1.75)
#krsa_top_hits(AvgZTable2, 1.5)

# Show the number of peptides per each set in atable
krsa_show_peptides(sigPeps_total)
```
B vs A


```{r}
krsa_zscores_plot(AvgZTable)
```
```{r}
krsa_zscores_plot(AvgZTable2)
```
Krsa reverse krsa plot
```{r}
krsa_reverse_krsa_plot(chipCov, diff_df, kinases_hits, 0.2, byChip = F)
```
krsa coverage plot


```{r}
krsa_coverage_plot(chipCov, AvgZTable2, chipType)
```
Ball Model Network
```{r}
krsa_ball_model(kinases_hits, AvgZTable2, 10, 2.5, 4.8)
```






```{r}
kinomedata<-read.csv("kinomedata.csv")
head(kinomedata, 10)
```














































